### PR TITLE
Gang bank

### LINKED ISSUE #190: Gang Funds Duping Money
Expected behaviour

- When 2 people withdraw money from the same gang fund, it should remove both amounts.

Actual behaviour

- When 2 people withdraw money from the same gang fund,at the same time it only removes the amount once, I have put a temporary fix in on my server where only the gang owner can access the gang funds, I think there is 2 options for a permanent fix.
1. remove the option to withdraw money  from the gang fund and make it an option to use the gang funds to purchase items instead of your own money.
2. put the gang bank on a different menu, maybe only accessible from rebel outposts, and only 1 member can access the gang funds at a time with set/get variable.

Steps to reproduce the behaviour

get 2 members in the same gang, have 1 million in the gang funds, count down and click remove x amount at the exact same time.

RPT and/or extDB2 logs (if applicable)
N/A

Mission version:
4.4R3 and 4.4R4

Launcher version:
N/A

Game version:
N/A

Branch:
N/A

### PR DESCRIPTION
**Pull request by DomT602**
Resolves #190 

#### Changes proposed in this pull request: 
- Fixes the above issue

- [x] I have tested my changes and corrected any errors found

### CODE REVIEW

File: Altis_Life.Altis/dialog/function/fn_useGangBank.sqf [hunk 1]
```diff
@@ -6,34 +6,34 @@
...
 
 //Series of stupid checks
 if (isNil {(group player) getVariable "gang_name"}) exitWith {hint localize "STR_ATM_NotInGang"}; // Checks if player isn't in a gang
 if (_value > 999999) exitWith {hint localize "STR_ATM_WithdrawMax";};
-if (_value < 0) exitWith {};
+if (_value < 1) exitWith {};
 if (!([str(_value)] call TON_fnc_isnumber)) exitWith {hint localize "STR_ATM_notnumeric"};
-if (_value > _gFund) exitWith {hint localize "STR_ATM_NotEnoughFundsG"};
+if ((_deposit && _value > CASH) || (!_deposit && _value > _gFund)) exitWith {hint localize "STR_ATM_NotEnoughCash"};
 if (_val < 100 && _gFund > 20000000) exitWith {hint localize "STR_ATM_WithdrawMin"}; //Temp fix for something.
```
Comment (BoGuu) [inline Altis_Life.Altis/dialog/function/fn_useGangBank.sqf]: `_val` is undefined. Perhaps we should also investigate briefly why this was ever needed? This number is above max precision regardless, perhaps that was the initial thought?

File: Altis_Life.Altis/dialog/function/fn_useGangBank.sqf [hunk 2]
```diff
@@ -6,34 +6,34 @@
...
-if ((group player getVariable ["gbank_in_use_by",player]) != player) exitWith {hint localize "STR_ATM_WithdrawInUseG"}; //Check if it's in use.
 
-_gFund = _gFund - _value;
-CASH = CASH + _value;
-group player setVariable ["gang_bank",_gFund,true];
+life_action_delay = time;
```
Comment (BoGuu) [inline Altis_Life.Altis/dialog/function/fn_useGangBank.sqf]: Would be better to have this as the very last statement; this function is running unscheduled

File: Altis_Life.Altis/dialog/function/fn_useGangBank.sqf [hunk 3]
```diff
@@ -6,34 +6,34 @@
...
+    [1,group player,_deposit,_value,player,CASH] remoteExecCall ["TON_fnc_updateGang",RSERV]; //Update the database.
 };
 
-hint format [localize "STR_ATM_WithdrawSuccessG",[_value] call life_fnc_numberText];
-[] call life_fnc_atmMenu;
-[6] call SOCK_fnc_updatePartial;
-
 if (LIFE_SETTINGS(getNumber,"player_moneyLog") isEqualTo 1) then {
     if (LIFE_SETTINGS(getNumber,"battlEye_friendlyLogging") isEqualTo 1) then {
         money_log = format [localize "STR_DL_ML_withdrewGang_BEF",_value,[_gFund] call life_fnc_numberText,[BANK] call life_fnc_numberText,[CASH] call life_fnc_numberText];
```
Comment (BoGuu) [inline Altis_Life.Altis/dialog/function/fn_useGangBank.sqf]: We could do money log stuff on the server w/ the initial call?

File: life_hc/MySQL/Gangs/fn_updateGang.sqf [hunk 4]
```diff
@@ -30,7 +31,33 @@ switch (_mode) do {
...
+        private _funds = _group getVariable ["gang_bank",0];
+        if (_deposit) then {
+            _funds = _funds + _value;
+            _group setVariable ["gang_bank",_funds,true];
+            [1,format["Successfully deposited $%1.",_value]] remoteExecCall ["life_fnc_broadcast",remoteExecutedOwner];
+            _cash = _cash - _value;
+        } else {
+            if (_value > _funds) exitWith {
+                [1,"There are not enough funds in the bank for this withdrawal."] remoteExecCall ["life_fnc_broadcast",remoteExecutedOwner];
+                breakTo "main";
```
Comment (BoGuu) [inline life_hc/MySQL/Gangs/fn_updateGang.sqf]: Would be more elegant to simply `breakOut ""`, and remove the scopeName above, this usage of breakTo is pretty spaghetti

File: life_server/Functions/Gangs/fn_updateGang.sqf [hunk 5]
```diff
@@ -28,7 +29,33 @@ switch (_mode) do {
...
+        private _funds = _group getVariable ["gang_bank",0];
+        if (_deposit) then {
+            _funds = _funds + _value;
+            _group setVariable ["gang_bank",_funds,true];
+            [1,format["Successfully deposited $%1.",_value]] remoteExecCall ["life_fnc_broadcast",remoteExecutedOwner];
+            _cash = _cash - _value;
+        } else {
+            if (_value > _funds) exitWith {
+                [1,"There are not enough funds in the bank for this withdrawal."] remoteExecCall ["life_fnc_broadcast",remoteExecutedOwner];
+                breakTo "main";
```
Comment (BoGuu) [inline life_server/Functions/Gangs/fn_updateGang.sqf]: Ditto

File: Altis_Life.Altis/dialog/function/fn_gangBankResponse.sqf [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+/*
+    File: fn_gankBankResponse.sqf
+    Author: DomT602
+    Description:
+    Receives response from the server.
+*/
+params [
+    ["_value",-1,[0]]
+];
+if (!isRemoteExecuted || remoteExecutedOwner != 2) exitWith {};
```
Comment (BoGuu) [inline Altis_Life.Altis/dialog/function/fn_gangBankResponse.sqf]: This isn't 2 for HC. Also last time I checked if the RE was from the HC it was simply broken. Might be worth checking before updating

**Comment by imthatguyhere** [comment]
> Other than that LGTM, does it work as expected? @imthatguyhere, did you test this already?

I have not. Not sure if @DomT602  has though. If not I can try to get some testing done on it in the next few days.

**Pull request merged**